File: student-service/src/main/java/com/mentorhub/studentservice/model/StudentProfile.java
package com.mentorhub.studentservice.model;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "student_profiles")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class StudentProfile {
    @Id
    @Column(name = "\"Id\"")
    private UUID id;

    @Column(name = "\"UserAccountId\"", nullable = false)
    private UUID userAccountId;

    @Column(name = "\"ProfilePhotoUrl\"")
    private String profilePhotoUrl;

    @Column(name = "\"DateOfBirth\"")
    private Instant dateOfBirth;

    @Column(name = "\"Gender\"", length = 40)
    private String gender;

    @Column(name = "\"CityOrArea\"", length = 160)
    private String cityOrArea;

    @Column(name = "\"EducationLevel\"", length = 120)
    private String educationLevel;

    @Column(name = "\"CurrentGradeOrYear\"", length = 120)
    private String currentGradeOrYear;

    @Column(name = "\"SchoolOrInstitutionName\"", length = 240)
    private String schoolOrInstitutionName;

    @Column(name = "\"MediumOfEducation\"", length = 60)
    private String mediumOfEducation;

    @Column(name = "\"SubjectsCsv\"")
    private String subjectsCsv;

    @Column(name = "\"TopicsOfDifficulty\"", length = 1200)
    private String topicsOfDifficulty;

    @Column(name = "\"TutoringPurpose\"", length = 500)
    private String tutoringPurpose;

    @Column(name = "\"LearningGoalsOrTargetGrade\"", length = 500)
    private String learningGoalsOrTargetGrade;

    @Column(name = "\"PreferredMode\"", length = 50)
    private String preferredMode;

    @Column(name = "\"PreferredDaysCsv\"")
    private String preferredDaysCsv;

    @Column(name = "\"PreferredTimeSlotsCsv\"")
    private String preferredTimeSlotsCsv;

    @Column(name = "\"BudgetPerSession\"")
    private BigDecimal budgetPerSession;

    @Column(name = "\"BudgetPerMonth\"")
    private BigDecimal budgetPerMonth;

    @Column(name = "\"PreferredTutorGender\"", length = 40)
    private String preferredTutorGender;

    @Column(name = "\"PreferredLanguageOfInstruction\"", length = 80)
    private String preferredLanguageOfInstruction;

    @Column(name = "\"GuardianFullName\"", length = 180)
    private String guardianFullName;

    @Column(name = "\"GuardianContactNumber\"", length = 40)
    private String guardianContactNumber;

    @Column(name = "\"GuardianEmailAddress\"", length = 180)
    private String guardianEmailAddress;

    @Column(name = "\"GuardianRelationship\"", length = 100)
    private String guardianRelationship;

    @Column(name = "\"GuardianConsentAcknowledgment\"")
    private Boolean guardianConsentAcknowledgment;

    @Column(name = "\"TermsAccepted\"")
    private Boolean termsAccepted;

    @Column(name = "\"PrivacyAccepted\"")
    private Boolean privacyAccepted;

    @Column(name = "\"ProfileCompleteness\"")
    private Integer profileCompleteness;

    @Column(name = "\"CreatedAtUtc\"", nullable = false)
    private Instant createdAtUtc;

    @Column(name = "\"UpdatedAtUtc\"")
    private Instant updatedAtUtc;
}

File: UserManagment/UserManagment.Domain/Entities/StudentProfile.cs
namespace UserManagment.Domain.Entities;

public class StudentProfile
{
    public Guid Id { get; set; }
    public Guid UserAccountId { get; set; }

    public string ProfilePhotoUrl { get; set; } = string.Empty;

    public DateTime? DateOfBirth { get; set; }
    public string? Gender { get; set; }
    public string CityOrArea { get; set; } = string.Empty;

    public string EducationLevel { get; set; } = string.Empty;
    public string CurrentGradeOrYear { get; set; } = string.Empty;
    public string? SchoolOrInstitutionName { get; set; }
    public string MediumOfEducation { get; set; } = string.Empty;

    public string SubjectsCsv { get; set; } = string.Empty;
    public string InterestsCsv { get; set; } = string.Empty;
    public string? TopicsOfDifficulty { get; set; }
    public string TutoringPurpose { get; set; } = string.Empty;
    public string? LearningGoalsOrTargetGrade { get; set; }

    public string PreferredMode { get; set; } = "Online";
    public string PreferredDaysCsv { get; set; } = string.Empty;
    public string PreferredTimeSlotsCsv { get; set; } = string.Empty;
    public decimal? BudgetPerSession { get; set; }
    public decimal? BudgetPerMonth { get; set; }
    public string? PreferredTutorGender { get; set; }
    public string PreferredLanguageOfInstruction { get; set; } = "English";

    public string? GuardianFullName { get; set; }
    public string? GuardianContactNumber { get; set; }
    public string? GuardianEmailAddress { get; set; }
    public string? GuardianRelationship { get; set; }
    public bool GuardianConsentAcknowledgment { get; set; }

    public bool TermsAccepted { get; set; }
    public bool PrivacyAccepted { get; set; }

    public int ProfileCompleteness { get; set; }
    public DateTime CreatedAtUtc { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAtUtc { get; set; } = DateTime.UtcNow;

    public UserAccount UserAccount { get; set; } = null!;

    public bool IsMinor()
    {
        if (!DateOfBirth.HasValue)
        {
            return false;
        }

        var today = DateTime.UtcNow.Date;
        var birthDate = DateOfBirth.Value.Date;
        var age = today.Year - birthDate.Year;
        if (birthDate > today.AddYears(-age))
        {
            age--;
        }

        return age < 18;
    }

    public int CalculateProfileCompleteness()
    {
        var checks = new List<bool>
        {
            !string.IsNullOrWhiteSpace(ProfilePhotoUrl),
            DateOfBirth.HasValue,
            !string.IsNullOrWhiteSpace(CityOrArea),
            !string.IsNullOrWhiteSpace(EducationLevel),
            !string.IsNullOrWhiteSpace(CurrentGradeOrYear),
            !string.IsNullOrWhiteSpace(MediumOfEducation),
            !string.IsNullOrWhiteSpace(SubjectsCsv),
            !string.IsNullOrWhiteSpace(InterestsCsv),
            !string.IsNullOrWhiteSpace(TutoringPurpose),
            !string.IsNullOrWhiteSpace(PreferredMode),
            !string.IsNullOrWhiteSpace(PreferredDaysCsv),
            !string.IsNullOrWhiteSpace(PreferredTimeSlotsCsv),
            !string.IsNullOrWhiteSpace(PreferredLanguageOfInstruction),
            TermsAccepted,
            PrivacyAccepted
        };

        if (IsMinor())
        {
            checks.Add(!string.IsNullOrWhiteSpace(GuardianFullName));
            checks.Add(!string.IsNullOrWhiteSpace(GuardianContactNumber));
            checks.Add(!string.IsNullOrWhiteSpace(GuardianRelationship));
            checks.Add(GuardianConsentAcknowledgment);
        }

        var completed = checks.Count(value => value);
        return (int)Math.Round((completed / (double)checks.Count) * 100, MidpointRounding.AwayFromZero);
    }
}
